server/raft_chain_of_blocks_helpers_test.go

```diff
@@ -0,0 +1,367 @@
+	s                          *Server
+	n                          RaftNode
+	cfg                        *RaftConfig
+	wg                         sync.WaitGroup
+	leader                     bool
+	proposalSequence           uint64
+	rng                        *rand.Rand
+	hash                       hash.Hash
+	blocksApplied              uint64
+	blocksAppliedSinceSnapshot uint64
+	stopped                    bool
+	ready                      bool
+}
+
+func (sm *RCOBStateMachine) waitGroup() *sync.WaitGroup {
+	sm.Lock()
+	defer sm.Unlock()
+	return &sm.wg
+}
+
+// RCOBBlock is the structure of values replicated through Raft.
+// Block data is a random array of bytes.
+// Additional proposer metadata is present for debugging and for further ordering invariant checks (but not included
+// in the hash).
+type RCOBBlock struct {
+	Proposer         string
+	ProposerSequence uint64
+	Data             []byte
+}
+
+// logDebug is for fine-grained event logging, useful for debugging but default off
+func (sm *RCOBStateMachine) logDebug(format string, args ...any) {
+	if RCOBOptions.verbose {
+		fmt.Printf("["+sm.s.Name()+" ("+sm.n.ID()+")] "+format+"\n", args...)
+	}
+}
+
+func (sm *RCOBStateMachine) server() *Server {
+	sm.Lock()
+	defer sm.Unlock()
+	return sm.s
+}
+
+func (sm *RCOBStateMachine) node() RaftNode {
+	sm.Lock()
+	defer sm.Unlock()
+	return sm.n
+}
+
+func (sm *RCOBStateMachine) propose(data []byte) {
+	sm.Lock()
+	defer sm.Unlock()
+	if !sm.ready {
+		sm.logDebug("Refusing to propose during recovery")
+	}
+	err := sm.n.ForwardProposal(data)
+	if err != nil {
+		sm.logDebug("block proposal error: %s", err)
+	}
+}
+
+func (sm *RCOBStateMachine) applyEntry(ce *CommittedEntry) {
+	sm.Lock()
+	defer sm.Unlock()
+	if ce == nil {
+		// A nil entry signals that the previous recovery backlog is over
+		sm.logDebug("Recovery complete")
+		sm.ready = true
+		return
+	}
+	sm.logDebug("Apply entries #%d (%d entries)", ce.Index, len(ce.Entries))
+	for _, entry := range ce.Entries {
+		if entry.Type == EntryNormal {
+			sm.applyBlock(entry.Data)
+		} else if entry.Type == EntrySnapshot {
+			sm.loadSnapshot(entry.Data)
+		} else {
+			panic(fmt.Sprintf("[%s] unknown entry type: %s", sm.s.Name(), entry.Type))
+		}
+	}
+	// Signal to the node that entries were applied
+	sm.n.Applied(ce.Index)
+}
+
+func (sm *RCOBStateMachine) leaderChange(isLeader bool) {
+	if sm.leader && !isLeader {
+		sm.logDebug("Leader change: no longer leader")
+	} else if sm.leader && isLeader {
+		sm.logDebug("Elected leader while already leader")
+	} else if !sm.leader && isLeader {
+		sm.logDebug("Leader change: i am leader")
+	} else {
+		sm.logDebug("Leader change")
+	}
+	sm.leader = isLeader
+	if isLeader != sm.node().Leader() {
+		sm.logDebug("⚠️ Leader state out of sync with underlying node")
+	}
+}
+
+func (sm *RCOBStateMachine) stop() {
+	sm.Lock()
+	defer sm.Unlock()
+	sm.n.Stop()
+	sm.wg.Wait()
+
+	// Clear state, on restart it will be recovered from snapshot or peers
```

Comment:
When running locally it seems the test can lock up and not do anything/exceed the test duration.

Should be fixed by not holding the lock while waiting for `Stop/Wait` to complete The same logic is used in `stateAdder.stop()`.
```suggestion
	n, wg := sm.node(), sm.waitGroup()
	n.Stop()
	n.WaitForStop()
	wg.Wait()

	// Clear state, on restart it will be recovered from snapshot or peers
	sm.Lock()
	defer sm.Unlock()
```